llvm/lib/MC/MCCASObjectV1.cpp

```diff
@@ -936,13 +937,83 @@ Error MCCASBuilder::createStringSection(
   return Error::success();
 }
 
+/// Reads and returns the length field of a debug info header contained in
+/// Reader, assuming Reader is positioned at the beginning of the section.
+/// The Reader's state is advanced to the first byte after the section.
+Expected<size_t> getSizeFromDebugHeaderAndSkip(BinaryStreamReader &Reader) {
+  // From Dwarf 5 section 7.4:
+  // In the 32-bit DWARF format, an initial length field [...] is an unsigned
+  // 4-byte integer (which must be less than 0xfffffff0);
+  uint32_t Word1;
+  if (auto E = Reader.readInteger(Word1))
+    return E;
+
+  // TODO: handle 64-bit DWARF format.
+  assert(Word1 < 0xfffffff0);
+
+  if (auto E = Reader.skip(Word1))
+    return E;
+  return Word1;
+}
+
+/// For each compile unit inside the __debug_info section, create a
+/// corresponding DebugInfoRef CAS object.
+Error MCCASBuilder::splitDebugInfoCUs() {
+  auto *DebugInfoSection =
+      Asm.getContext().getObjectFileInfo()->getDwarfInfoSection();
+  if (!DebugInfoSection)
+    return Error::success();
+
+  const auto &FragmentList = DebugInfoSection->getFragmentList();
+  assert(FragmentList.size() == 1 && "One fragment in debug str section");
```

Comment:
Felipe, this assert is not needed, we do not care if we have one debug info fragment or more than one, also your assert says "debug str section"

Reply:
I see, that makes sense. At some point we should find a strategy to both account for multiple fragments _and_ have a zero-copy strategy here. Creating a vector and copying the entire debug section will be expensive in the long run, but for now I guess this is good enough.

Will change!